### PR TITLE
component TypeScript support

### LINKED ISSUE #130: Add typescript compilation step to build
It would be nice to be able to use typescript for routers, etc. This should also make it easier to share code in the client files (search and content type dashboards).

Ideally, we would output the build files to a directory like `target`, `build`, or `dist` but since the configuration for webgme is automatically generated, we should keep the final files in `src` and move the src files to something like `lib/`. Alternatively, we could keep the built files in the same directory (so both are in `src` and they are just alongside their ts source files).

### PR DESCRIPTION
**Pull request by romancow**
Setting up some basic TypeScript support for custom webgme components.

I converted the existing "Dashboard" router component as a test / example.

One plus to the approach I took is that it allows some degree of customization for how a component using TypeScript is built, with a tradeoff being there is a small bit of setup for it.

The main key when adding a component with Typescript, is to first add a new [`tsconfig.json`](https://www.typescriptlang.org/tsconfig) to your component's folder that extends the base config and specifies what files need to be built (via the [`files`](https://www.typescriptlang.org/tsconfig#files) or [`include`](https://www.typescriptlang.org/tsconfig#include) values):
```json
{
  "extends": "../../tsconfig.component.json",
  "include": [ "*.ts"]
}
```
This allows you to customize exactly what files will be built, along with really any other specific build options you want to override.

Next, add a reference to this `tsconfig.json` in the main `tsconfig.json` file at the root of the project. That way when the entire project is built, your component will be included.

That's pretty much the essentials, but there are a few other things to keep in mind. Under this approach, you could technically keep your Typescript and built files alongside one another in the webgme "./src" folder (such as I did with the existing "Dashboard" router) --***or***-- put your typescript source somewhere else as long as the reference in the main `tsconfig.json` points to it and you override your `outDir` to put your build files in the proper location. Either way, you probably want to make sure your built files are ignored by git. This could be by adding a `.gitignore` for it in the component's corresponding `./src` folder, or by outputting the build to a folder that is already being ignored (such as I did with the "Dashboard" router by outputting to a `build` folder). If you take the latter approach, also be sure to change corresponding config values (such as in `./config/config.webgme.js`) to match this new location.

There are some other advantages to leveraging [TypeScript's Project References](https://www.typescriptlang.org/docs/handbook/project-references.html), such as incremental build and the ability to delete all the outputs with a simple `--clean` flag.

resolves #130

### CODE REVIEW

**Comment by romancow** [comment]
A quick example on adding a simple one file component (say a router):

First run the webgme cli command: `webgme new router MyNewRouter`

Next, add a `./src/routers/MyNewRouter/tsconfig.json` file:
```json
{
  "extends": "../../tsconfig.component.json",
  "files": [ "MyNewRouter.ts"]
}
```

Then add a reference to this in `./tsconfig.json`:
```json
{
  "files": [],
  "references": [
    { "path": "src/routers/Dashboard" },
    { "path": "src/routers/MyNewRouter" },  // new line added here
  ]
}
```

Change the file extension of your plugin from `.js` to `.ts`:
 `src/routers/MyNewRouter/MyNewRouter.js` to  `src/routers/MyNewRouter/MyNewRouter.ts`
You can then delete code in this file or use it as a starting point for your TypeScript version.

Finally, add a `src/plugins/MyNewRouter/.gitignore` file to ignore the built output:
```gitignore
MyNewRouter.js
```

File: config/config.webgme.js [hunk 1]
```diff
@@ -38,7 +38,7 @@ config.rest.components['JSONSchema'] = {
   options: {}
 };
 config.rest.components['Dashboard'] = {
-  src: __dirname + '/../src/routers/Dashboard/Dashboard.js',
+  src: __dirname + '/../src/routers/Dashboard/build/',
```
Comment (brollb) [inline config/config.webgme.js:41]: We will need to change this bc the `config.webgme.js` file is programmatically generated and will lose this edit as soon as another plugin/router/etc is created. To test, try running `webgme refresh` in the project root.

**Review by brollb — CHANGES_REQUESTED** [review]
I left a comment below blocking the merge that should probably result in a new feature being added to webgme-cli to support components with a build step.

Ideally, it would be nice if we could support another key in the components entry in the webgme-setup.json file. Probably something like `dist` which could reference the built file to use in the config. If it doesn't exist, it can default to the current approach.

File: config/config.webgme.js (hunk 1 above)
Comment (romancow) [reply config/config.webgme.js]: Alright, this should be resolved now.

**Review by brollb — APPROVED** [review]

**Pull request merged**